32 update model attributes

emrdagkusu opened the pull request:
This merge request consists of the followings:

- Model review and restructuring models
- Unit / integration tests
- FastAPI and Pydantic validation tricks for models 
- New validations and tests
- Updating university section logic, attendance logic

closes #10 #14 #21 #29 #30 #31 #32

emrdagkusu committed:
add changes of user and university models

emrdagkusu committed:
update university models and lesson section logic

emrdagkusu committed:
update semester start hour

emrdagkusu committed:
update section logic from integer list to model

emrdagkusu committed:
remove id from section slots

emrdagkusu committed:
update slot and absence logic

put absences into slots to be able to relate absence with slots

emrdagkusu committed:
sort absences while adding

emrdagkusu committed:
update slot order while adding/updating slots

emrdagkusu committed:
update order of sections and slots while adding/updating sections

emrdagkusu committed:
do not let creating new university lesson with same code

emrdagkusu committed:
add new attriubutes and routers to user and semester

emrdagkusu committed:
add router for updating universities with new attributes

emrdagkusu committed:
update model for admin user

emrdagkusu committed:
update university tests after changing model attributes

shiny-apricot committed:
add user/semester and user/user router tests

shiny-apricot committed:
Add user/lesson tests.

shiny-apricot committed:
Add user/slot_routers tests.

emrdagkusu committed:
let empty room for slots

emrdagkusu committed:
Generate openapi.json

shiny-apricot commented on backend/app/models/uni_models.py:
```diff
@@ -226,35 +223,90 @@ class Config:
 
-class UpdateUniversityNameModel(BaseModel):
-    name: str = Field(...)
+class UpdateUniversityModel(BaseModel):
```
We may need to left some university_update_model fields Optional.

emrdagkusu replied:
If we want them to be optional, we should find a way for MongoDB to update few of them or we can iterate through the attributes of the model and if it is not none we can update it. Which one will be better?

emrdagkusu replied:
If we want them to be optional, we should find a way for MongoDB to update few of them or we can iterate through the attributes of the model and if it is not none we can update it. Which one will be better?

emrdagkusu replied:
```
 class UpdateUniversityModel(BaseModel):
    name: Optional[str] = Field(min_length=2, max_length=100)
    website: Optional[str] = Field(min_length=5, max_length=255)
    country: Optional[str] = Field(min_length=2, max_length=100)
    city: Optional[str] = Field(min_length=2, max_length=100)
    address: Optional[str] = Field(min_length=2, max_length=100)
    phone: Optional[str] = Field(min_length=2, max_length=100)
    email: Optional[str] = Field(min_length=2, max_length=100)
    zip_code: Optional[str] = Field(min_length=2, max_length=100)
    description: Optional[str] = Field(max_length=255)
    logo: Optional[str] = Field(max_length=255)
    cover_photo: Optional[str] = Field(max_length=255)

    class Config:
        schema_extra = {
            "example": {
                "name": "Test University",
                "website": "test.com",
                "country": "USA",
                "city": "Los Angeles",
                "address": "123 Main Street",
                "phone": "555-555-5555",
                "email": "info@test.com",
                "zip_code": "90210",
                "description": "Test University Description",
                "logo": "test_logo.png",
                "cover_photo": "test_cover_photo.png",
            }
        }
```

### INSIDE UPDATE ROUTER

```
updated_features = {}
for key in university:
       if university[key] is not None:
            updated_features.update({key: university[key]})

update_result = await request.app.mongodb["universities"].update_one(
       {"_id": unid}, {"$set": updated_features}
)
```

This can be a solution to partially updating models. In the update_XXX models we can make everything Optional and then we can update not None fields in the update router. I can add this feature to other update endpoints as well.

shiny-apricot commented on backend/app/models/uni_models.py:
```diff
@@ -118,105 +126,94 @@ class Config:
+        schema_extra = {"example": {"name": "COMP",}}
 
 
 class UniversityModel(BaseModel):
```
We may need to left some uni_model fields Optional.

emrdagkusu replied:
If we want them to be optional, we should find a way for MongoDB to update few of them or we can iterate through the attributes of the model and if it is not none we can update it. Which one will be better?

emrdagkusu replied:
Same with upper case.

emrdagkusu replied:
Same solution for above can be used here.

emrdagkusu committed:
let fields in update models optional

in update model, fields are optional anymore. thus, it is not required to give old values to update a model

emrdagkusu committed:
Merge branch '32-update-model-attributes' of github.com:hbusul/kucukdev_api into 32-update-model-attributes

emrdagkusu committed:
Generate openapi.json

shiny-apricot reviewed (approved):
Seems great!
